### ISSUE TITLE
AN IDEA!

### ISSUE THREAD

**Issue by quitequinn** [weiweihuanghuang/GaugeTool#1]
Instead of having a fixed oval as the form to determine stroke width why not calculate the shortest straight line between any two points of your paths and the cursor. That way it could live calculate the CURRENT stroke width and you could use the tool to determine transitions of any stroke width? (not just monolinear ones)

**Comment by quitequinn** [comment]
For example:
![screen shot 2016-07-04 at 3 43 52 pm](https://cloud.githubusercontent.com/assets/2885870/16570628/2710b0a4-41fe-11e6-9f49-798cc273334a.jpg)

**Comment by weiweihuanghuang** [comment]
Great Idea! I actually originally wanted to make the ball sticky and stick to a certain side to achieve a more accurate reading, but that was alas beyond my current abilities... This is even harder I believe, especially to find where the line should be drawn. I suggested to @simoncozens if his [Calipers](https://github.com/simoncozens/Callipers) tool could work around an entire /o and I think it was a similar issue https://twitter.com/w__h_/status/710460919151763456

**Comment by quitequinn** [comment]
RIGHT THE OPPOSING SIDE.. and the opposite path too. 

_Lets explore that for a second. Lets do it for an O(two paths) and an S(one path)._
As long as the goal is to find a strait line I don't think this is a problem. Even if we compare EVERY path... as long as it only shows the closest path we should be set. Here is a possible solution (please forgive my made up functions and javascript.. I don't think as fluidly in python):

```javascript
var smallestDistance;
var lines = [];

var glyph = { 
   name: "S",
   paths: [ ],
};
foreach path, pathIndex in glyphFromRobofont {
   glyph.paths.push({
      index: pathIndex,
      name: path.name,
      instances: []
   });
   foreach instance, instanceIndex in path.instances {
      glyph.paths[pathIndex].push({
         index: instanceIndex,
         name: instance.name,
         distanceFromCursor: distance(instance.location, cursor.location)
   }
   foreach instance, instanceIndex in glyph.paths.instances {
      foreach instance2, instanceIndex2 in glyph.paths.instances {
         if (line(instance, cursor.location, instance2)){
            line.push({
               instance1: instance,
               cusor: cursor,
               instance2: instance2,
               distance: distance(instance.location, cursor.location, instance2.location)
            });
         }
      }
   }
}


smallestDistance = smallestInList(lines, distance);

// This is the line that is the shortest length that intersects any path and your cursor
style(smallestDistance);
```


Before I wrote that solution I considered an alternative... I don't think this conversation is helpfull in the case of a trait line but maybe it is useful for Gauge tool. 

**How do we find the opposite path:** This will only happen with two or more paths so we don't need to explore S. Why not just calculate distances for all paths? BUT then there is the possibility of having overlapped measurements.. but that probably wouldn't be the end of the world. 

**How do we find the opposing side:** Well you could calculate all instance of a path. You could then display those instances in a list with the instance id and distances from that instances' location and the cursor. Once you have that all you need to do is divide that list in half and you should have a range that doesn't include the other side of a glyph.

**Comment by simoncozens** [comment]
Unfortunately, just finding the closest segment doesn't work. The closest segment from the marked point of this bold /n is not the other side of the stem, but the other side of the aperture:

![screen shot 2016-07-08 at 13 01 29](https://cloud.githubusercontent.com/assets/106728/16676292/489d5f22-450c-11e6-9bfc-251b765ac5b8.png)

**Comment by quitequinn** [comment]
Right! but if you put the cursor within the body of the "n" it would draw a line to the other side of the stem.  If your cursor was within the counter of the "n" wouldn't you expect it to draw the line within the cursor?

**Comment by quitequinn** [comment]
if you wanted to get super specific you could add an if that doesn't draw any line if the cursor is actually on a path.

**Comment by quitequinn** [comment]
ALTHOUGH you could have situations like so... but no tool is perfect.

![screen shot 2016-07-07 at 8 10 35 pm](https://cloud.githubusercontent.com/assets/2885870/16676426/e268611c-447e-11e6-94b9-a5615ba06aa4.jpg)

**Comment by quitequinn** [comment]
Some more examples..
![screen shot 2016-07-07 at 8 12 49 pm](https://cloud.githubusercontent.com/assets/2885870/16676467/380400cc-447f-11e6-9095-9e1d44feb3a6.jpg)
![screen shot 2016-07-07 at 8 12 14 pm](https://cloud.githubusercontent.com/assets/2885870/16676468/3817f64a-447f-11e6-8108-5fbf32d0cc19.jpg)
![screen shot 2016-07-07 at 8 11 43 pm](https://cloud.githubusercontent.com/assets/2885870/16676469/381dc8cc-447f-11e6-8b6c-fafed710e74e.jpg)

**Comment by simoncozens** [comment]
Incidentally, Fontlab VI does this, and it's really funky:
![flvi-stem](https://cloud.githubusercontent.com/assets/106728/16676525/41562b48-450e-11e6-87a6-97ed69385c97.gif)

**Comment by quitequinn** [comment]
Right! That's the same idea but using the heart line instead of the cursor.

**Comment by weiweihuanghuang** [comment]
Do you know what though, this would be a GREAT tool for measuring the width of diagonal stems, I don't know the right equations to find the two closest point between two segments.

**Comment by simoncozens** [comment]
Glyphs has Objective C functions you can use to do this. I don't know if they're exposed to Python. Look in GSGeometrieHelper.h
